optaplanner-core/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java

```diff
@@ -238,110 +237,70 @@ public void skipAhead() throws ExecutionException, InterruptedException {
-                                                                                        if (bestSolution.getEntityList().get(2).getValue() == null) {
-                                                                                            try {
-                                                                                                latch.await();
-                                                                                            } catch (InterruptedException e) {
-                                                                                                fail("Latch failed.");
-                                                                                            }
-                                                                                        } else if (bestSolution.getEntityList().get(3).getValue() == null) {
-                                                                                            fail("No skip ahead occurred: both e2 and e3 are null in a best solution event.");
-                                                                                        }
-                                                                                    });
+                problemId -> PlannerTestUtils.generateTestdataSolution("s1", 4),
+                bestSolution -> {
+                    if (bestSolution.getEntityList().get(1).getValue() == null) {
+                        // The problem itself causes a best solution event. TODO Do we really want that behavior?
+                        return;
+                    }
+                    eventCount.incrementAndGet();
+                    if (bestSolution.getEntityList().get(2).getValue() == null) {
+                        try {
+                            latch.await();
+                        } catch (InterruptedException e) {
+                            fail("Latch failed.");
+                        }
+                    } else if (bestSolution.getEntityList().get(3).getValue() == null) {
+                        fail("No skip ahead occurred: both e2 and e3 are null in a best solution event.");
+                    }
+                });
         assertSolutionInitialized(solverJob1.getFinalBestSolution());
         // EventCount can be 2 or 3, depending on the race, but it can never be 4.
         assertTrue(eventCount.get() < 4);
     }
 
-    /**
-     * In order to effectively test the terminateEarly() and close() methods there had to be a way how to make the job
-     * status change in a deterministic way while solving jobs on separate threads. The test waits for a status change in
-     * a loop and time-outs when it doesn't reach it's expected status.
-     */
-    @Ignore("https://issues.redhat.com/browse/PLANNER-1836")
-    @Test
-    public void terminateEarlyAndClose() throws InterruptedException {
+    @Test(timeout = 600_000L)
+    public void terminateEarly() throws InterruptedException {
```

Comment:
@michaltomco An attempt at trying to explain how I changed this test:
- The original test uses waitForJobStatusChangeOrTimeout() - but these are race condition sensitive. For example, presume we wait for SOLVING_ACTIVE. In theory, it's possible that the solver runs so quick, that between polling it goes from SOLVING_SCHEDULED to SOLVING_ACTIVE to NOT_SOLVING.
- Furthermore - the TimeUnit.MILLISECONDS.sleep(5) could be both too quick (see "live lock" in Brian Goetz's book) and too slow (increases total unit testing time, which is a cost for CI and all developers for every commit). Whenever possible, avoiding it is nice.

**The latches approach follows a differently philosophy.** Instead of trying to "poll" at the right time (retrying if needed) to confirm the sequence of events happens in the order we except them to happen, we pause the sequence of events with latches, at which point we can guarantee the solving status.

Reply:
Thanks for the explanation Geoffrey, I agree it is safer this way after reading your arguments.